File: app/src/main/java/com/example/my_board/ListViewAdapter.java
package com.example.my_board;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import org.w3c.dom.Text;

import java.util.ArrayList;
import java.util.HashMap;

public class ListViewAdapter extends BaseAdapter {
    private ImageView board_icon;
    private TextView board_title;
    private TextView board_uid;
    private TextView countLike;
    private String uid;
    private String title;
    private int isboard;
    private int like;

    public String getLike() {
        return Integer.toString(like);
    }

    public void setLike(String like) {
        this.like = Integer.parseInt(like);
    }

    private ArrayList<ListViewItem> listViewItemList = new ArrayList<ListViewItem>();

    public ListViewAdapter(){}

    @Override
    public int getCount(){
        return listViewItemList.size();
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent){
        final int pos = position;
        final Context context = parent.getContext();
        final ListViewItem listViewItem = listViewItemList.get(position);
        final View[] finalConvertView = {convertView};
        final ViewGroup finalParent = parent;

        if(convertView == null){
            LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = inflater.inflate(R.layout.listview_item,parent,false);
        }

        board_icon = (ImageView) convertView.findViewById(R.id.board_img);
        board_title = (TextView) convertView.findViewById(R.id.board_title);
        board_uid = (TextView)convertView.findViewById(R.id.board_uid);
        countLike = (TextView)convertView.findViewById(R.id.countLike);

        board_icon.setImageDrawable(listViewItem.getBoard_icon());
        board_title.setText(listViewItem.getBoard_title());
        board_uid.setText(listViewItem.getBoard_uid());
        countLike.setText(listViewItem.getCountLike());

        return convertView;
    }

    @Override
    public long getItemId(int position){
        return position;
    }

    @Override
    public Object getItem(int position){
        return listViewItemList.get(position);
    }

    public void addItem(Drawable icon, String title, String content, String uid, String countLike){
        ListViewItem item = new ListViewItem(icon, title, uid);
        item.setBoard_content(content);
        item.setCountLike(countLike);
        listViewItemList.add(item);

    }

    public void clear() {
        listViewItemList.clear();
    }

    public void setUId(String uid){
        this.uid = uid;
    }

    public void setBoard_title(String title){
        this.title = title;
    }

    public void setIsboard(int board){
        this.isboard = board;
    }

}

File: app/src/main/java/com/example/my_board/ListView/ListViewAdapter.kt
package com.example.my_board.ListView
import android.content.Context
import android.graphics.drawable.BitmapDrawable
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.content.res.ResourcesCompat
import com.example.my_board.R
import java.util.*


class ListViewAdapter : BaseAdapter(), Filterable {
    private var board_icon: ImageView? = null
    private var listFiter : Filter? = null
    private var board_title: TextView? = null
    private var board_uid: TextView? = null
    private var countLike: TextView? = null
    private var isboard = 0
    private var listViewItemList = ArrayList<ListViewItem>()
    private var filteredItemList = listViewItemList

    override fun getCount(): Int {
        return filteredItemList.size
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        var convertView = convertView
        val context = parent?.context
        val listViewItem = filteredItemList[position!!]

        if (convertView == null) {
            val inflater = context?.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            convertView = inflater.inflate(R.layout.listview_item, parent, false)
        }
        if(convertView != null){
            board_icon = convertView.findViewById<View>(R.id.character_image) as ImageView
            board_title = convertView.findViewById<View>(R.id.board_title) as TextView
            board_uid = convertView.findViewById<View>(R.id.board_uid) as TextView
            countLike = convertView.findViewById<View>(R.id.countLike) as TextView
            board_icon!!.setImageDrawable(listViewItem.character)
            board_title!!.text = listViewItem.board_title
            board_uid!!.text = listViewItem.board_uid
            countLike!!.text = listViewItem.countLike

        }
        return convertView!!
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItem(position: Int): Any {
        return filteredItemList[position]
    }

    fun addItem(title: String?, content: String?, uid: String?, countLike: String?, icon: BitmapDrawable) {
        val item = ListViewItem(title!!, uid!!, icon)
        item.board_content = content
        if(countLike != null){
            item.countLike = countLike!!
        }else{
            item.countLike = ""
        }
        listViewItemList.add(item)
    }

    fun clear() {
        filteredItemList.clear()
    }

    fun setIsboard(board: Int) {
        isboard = board
    }

    override fun getFilter(): Filter {
        if(listFiter == null){
            listFiter = ListFilter()
        }
        return listFiter as Filter
    }

    inner class ListFilter : Filter(){
        private val results = FilterResults()
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            if(constraint == null || constraint.length == 0){
                results.values = listViewItemList
                results.count = listViewItemList.size
            }else{
                val itemList = ArrayList<ListViewItem>()
                for(item in listViewItemList){
                    if (item.board_uid!!.toUpperCase().contains(constraint.toString().toUpperCase()) ||
                            item.board_title!!.toUpperCase().contains(constraint.toString().toUpperCase()))
                    {
                        itemList.add(item) ;
                    }
                }
                results.values = itemList
                results.count = itemList.size
            }
            return results
        }

        override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
            filteredItemList = results!!.values as ArrayList<ListViewItem>
            if(results.count > 0){
                notifyDataSetChanged()
            }else{
                notifyDataSetInvalidated()
            }
        }
    }

}

